cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,297 @@
+        return score;
+      } else {
+        return null;
+      }
+    }
+  }
+
+  private boolean conditionalReleaseAgent(Agent agent, String acquireScore) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      if (jedis.evalsha(getScriptSha(CONDITIONAL_SWAP_SET_SCRIPT),
+          Arrays.asList(WORKING_SET, WAITING_SET),
+          Arrays.asList(agent.getAgentType(), score(intervalProvider.getInterval(agent).getInterval()),
+              acquireScore)) != null) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  private boolean releaseAgent(Agent agent) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      if (jedis.evalsha(getScriptSha(SWAP_SET_SCRIPT),
+          Arrays.asList(WORKING_SET, WAITING_SET),
+          Arrays.asList(agent.getAgentType(), score(intervalProvider.getInterval(agent).getInterval()))) != null) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  private void saturatePool() {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      // Occasionally repopulate the agents in case redis went down. If they already exist, this is a NOOP
+      if (runCount % REDIS_REFRESH_PERIOD == 0) {
+        for (String agent : agents.keySet()) {
+          jedis.evalsha(getScriptSha(ADD_AGENT_SCRIPT), 2, WAITING_SET, WORKING_SET, agent, score(NOW));
+        }
+      }
+
+      // First cull threads in the WORKING set that have been there too long (TIMEOUT time).
+      Set<String> oldKeys = jedis.zrangeByScore(WORKING_SET, "-inf", score(NOW));
+      for (String key : oldKeys) {
+        // Ignore result, since if this agent was released between now and the above jedis call, our work was done
+        // for us.
+        AgentWorker worker = agents.get(key);
+        if (worker != null) {
+          releaseAgent(worker.agent);
+        }
+      }
+
+      // Now look for agents that have been in the queue for at least INTERVAL time.
+      List<String> keys = new ArrayList<>();
+      keys.addAll(jedis.zrangeByScore(WAITING_SET, "-inf", score(NOW)));
+      Set<AgentWorker> workers = new HashSet<>();
+
+      // Loop until we either run out of threads to use, or agents (which are keys) to run.
+      while (!keys.isEmpty() && runningAgents.map(Semaphore::tryAcquire).orElse(true)) {
+        String agent = keys.remove(0);
+
+        AgentWorker worker = agents.get(agent);
+        String score;
+        if (worker != null && (score = acquireAgent(worker.agent)) != null) {
+          // This score is used to determine if the worker thread running the agent is allowed to store its results.
+          // If on release of this agent, the scores don't match, this agent was rescheduled by a separate thread.
+          worker.setScore(score);
+          workers.add(worker);
+        }
+      }
+
+      for (AgentWorker worker : workers) {
+        Future s = agentWorkPool.submit(worker);
+        assert !s.isCancelled(); // This is only to appease findBugs
```

Comment:
there is an `@SuppressFBWarnings` annotation you could use instead of the assert here